
# Introduction
The aim of this notebook purports to illucidate the flow of nutrients through an ecosystem.
This will be done by taking an engineering centered approach. Despite the fact that the chemistry and bacteria involved in the nitrogen cycle are well studied, there is a lot to be said regarding the mass and energy fluxes and the varying amounts of the species that influence this sensitive flows of material and energy. With the former in mind, it is the aim of this notebook to set forward a general flux model that intends to illustrate how mass and energy flows through an [ecosystem](https://www.boredpanda.com/sealed-bottle-garden-david-latimer/?utm_source=google&utm_medium=organic&utm_campaign=organic).

![ecosystem](ecosystem_introduction.png)

#  Approach

Initially, it will be advantageous to select an aquatic ecosystem, as it hold the following advantages:
* Uniform concentrations throughout the sample volume;
* All the organisms can be selected carefully to be either single-celled or to possess uniform structures; and
* The availability of data pertaining to the abovementioned microorganisms.

In summary, a few of the microbes in the system will be used to produce simple flux models, which will expand both in amount and complexity over the course of the project. These flux models will result in algebraic response functions and their bio-masses will yield differential equations,all of which will be used to study the flow of nutrients through the ecosystem







# Plant model 





The reactions taking place in the plant will be sperated into two diferent models, 1: Photosynthesis and 2: Growth model

Photosynthesis is the process by which plants, algae, and certain bacteria (such as cyanobacteria) harnass energy from sunlight and convert it into chemical energy. During oxygenic photosynthesis light energy is captured and fascilitates the transfer of electrons from water to carbon dioxide to produce carbohydrates such as sugar. Therefore carbon dioxide is reduced and water is oxidized. Ultimately oxygen gas is produced alongside the sugar in the reaction as set out below:

$$CO_{2} + H_{2}O + light → CH_{2}O + O_{2}$$

In turn, the sugar produced will provide the energy to support energy production for plant growth. The proposed model for the synthesis of [biomass](https://www.researchgate.net/publication/223835531_An_overview_of_the_chemical_composition_of_biomass) can be set out as follow:

$$aCH_{2}O + bO_{2} → CH_{1.5}O_{0.64}N_{0.04} + dCO_{2} + eH_{2}O  $$

![Plant_Model](Plant_Modell.jpg)

Lastly, for simplicity it is assumed that the plant only absorbs $NO_{3}^{-}$ as the source of nitrogen and that $NH_{4}^{+}$ and other forms of nitrogen will not be absorbed by the plant.

# Nitrogen cycle





There are numerous types of bacteria that play their part in the nitrogen cycle. Therefore, it was thought good to name but a few of the types of bacteria that will play a part in the ecosystem. The index of bacteria provided is by no means exhaustive and will serve to simplify the modelling of the two phase system.

![index](Bacteria_Index.jpg)

The nitrogen cycle can be summarised by the  diagram below. Each individual step of our model will be discussed accordingly.

![index](Nitrogen_Cyclle.jpg)









## Nitrogen fixation

Nitrogen plays a big role in a plant growth as it forms part of the cholorphyll and the ammino acids which make the proteins of which the plant consists. Quite ironically, despite 79% of the air consisting of molecular nitrogen $N_{2}$ plants are unable to assimilate it due it being inert. So... for the plants to obtain nitrogen it has to be converted into forms that are susceptible to uptake by the plant such as Ammonium ($NH_{4}^+$) or Nitrate ($NO_3^{-}$) so that they can produce DNA, RNA, and proteins which are require nitrogen. This process is called nitrogen fixation and is performed by heterocysts, which are specialized nitrogen fixing cells in filamentous cyanobacteria [filamentous nitrogen bacteria](). Looking at the image below it is evident that the formation of heterocysts inside photosynthetic cells is known as cooperation, where a symbiotic sharing of metabolic products take place. For instance: the heretocysts share the ammonia produced and the photosynthetic green cells produce and share glucose.

![Heterocyst](Heterocyst.jpg)

Some cyanobacteria have a compartmentilisation strategy due to the fact that nitrogen fixation relies on an enzyme called nitrogenase which is inactivated by oxygen. Therefore, if you have photsynthesis occcuring, you cannot have nitrogen fixation occuring. Logically, the heterocysts have to create an anaerobic enviroment that is seperated from the photosynthetic cells so that fixation can take place.

Nitrogen fixation as a [chemical reaction](https://www.sciencedirect.com/science/article/pii/B9780128146675000088) can be described as follow:

$$ N_2 + 10H^{+} + 8e^{-} + 16MgATP  → 2NH_{3} + H_{2} + 16MgADP + 16Pi $$

For the ecosystem chosen we have chosen to use a simplified nitrogen fixing model. We will assume that the growth rate of the nitrate fixing bacteria will be related to the uptake of glucose and nitrogen , i.e.
$$ X(CH_{2}O,N_{2}) $$



Nitrogeen uptake rate given as $24.75$ $\mu mol.L^{-1}.h^{-1}$ [LOOK AT TABLE 7 page 286](https://mmbr.asm.org/content/mmbr/49/3/270.full.pdf)


With $μ_{2}$ reported as $0.085 ± 0.015 h^{-1}$ and the assumption that the nitrogen fixation is the only mechanism by which the bacteria is assimilating nitrogen. Nitrogen uptake has reported to be between $ 0.13 - 0.17 $ $nmol$ $C_{2}H_{4}$ $.mg$ $dry$ $weight^{-1}. h^{-1}$. With these parameters known, the model below will be sufficient to play its part in the ecosystem.

The following image is represents nitrogen fixation used in the model for the system.
![X2](Nitrogen_fixating_bacteria_explainedd.jpg)


In [1]:
import numpy as np
Stoichiometry_Matrix = np.matrix([[-1, 0, 1, 0, 1],
                        [0, -2, 0.2, 1, 0],
                        [-4, 6, 4.4, 1, 0],
                        [1, 0, 0, 0, 0],
                        [0, 0, 1, 0, 0]])
sol = np.matrix([[0, 0, 0, 1, 0.2]]).T
knownrates = np.linalg.solve(Stoichiometry_Matrix,sol) 
knownrates
#R_S, R_N2, RX, RNH4, RCO2

matrix([[1.   ],
        [0.395],
        [0.2  ],
        [0.75 ],
        [0.8  ]])

# Nitrification

Can be described as the microbially fascilitated two step biological process that fascilitates the  oxidation of ammonia to nitrate.Therefore, multiple types of bacteria are needed to perform the process of nitrification such as ammonia oxiding bacteria and nitrite oxidizing bacteria.

In order to simplify the model, the nitrifying bacteria model will assume that the bacteria chosen will be able to do both the oxidation of ammonia to nitrite and the oxidation of nitrite to nitrate. Therefore, the initial model will be something like:

![Nitrification_Steps](Nitrification_Steps.jpg) 

Furthermore,similar $\mu$ values and $NH_{4}^{+}$ $/$  $NO_{2}$ uptake rates will be assumed. Subsequently, the model will assume that growth is a function of the rate of uptake of nitrogen uptake, with a reported specific growth rate:
$$X_{3}(N-Uptake)$$

Literature has confirmed that a typical biomass formula for nitrifying bacteria is in the order of $CH_{1.4}N_{0.2}O_{0.4}$.Furthermore, Nitrifying bacteria make use of $CO_{2}$ as their primary source of carbon and make use of either $NH_3$$/$$NH_{4}^{+}$ or $NO_{2}$ to produce ATP. The [overall reaction](https://www.sciencedirect.com/science/article/pii/B9780124701007500224) for the production of new biomass coupled with ammonia can be represented as follow:

$$ NH_{3} + 1.83 O_{2} + 1.98 HCO_{3}^{-1}  →  0.021 C_{5}H_{7}NO_{2}
+ 1.041 H_{2}O + 0.98 NO_{3} + 1.88 H_{2}CO_{3}$$

The  physilogical parameters for nitrifying bacteria growth kinetics were obtained with a [specific growth](https://www.researchgate.net/publication/247876208_Growth_kinetic_studies_of_nitrifying_bacteria_by_the_immunofluorescent_counting_method) $\mu_{3} = 0.81$ $day^{-1}$

![Nitrification_bacteria_explained](Nitrification_bacteria_explained.jpg)


## Denitrification

Denitrification can be regarded as the micribially fascilitated process where nitrates are reduced to gaseous nitrogen, mostly, $N_{2}O$ and $N_{2}$. Denitrification is a response to a drop in oxygen concentration, which enables the anaerobic use of $NO_{2}$. Naturally in ecosystems this process is beneficial for the removal of nitrates. The bacterial growth model will be a function of nitrate uptake rate i.e. 
$$X(NO_{3}^{-1})$$


[overallreaction](https://sci-hub.tw/https://doi.org/10.1016/S0166-1116(08)70526-0) using CO2 as source of carbon...
id glucose is the electron acceptor then denitrification can be written as [Kinetics of Nitrification and Denitrification
Reactions](https://sci-hub.tw/10.1002/bit.260200105)



[insertbacteria formula,parracoccus Denitrificans](https://www.sciencedirect.com/science/article/pii/S2405844018385852#bib48)


![X3](Denitryfying_bacteria_explainedd.jpg)

$\mu$ (muvalues)[https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6936282/]


## [Decomposer](https://www.education.txstate.edu/ci/faculty/dickinson/PBI/PBISpring05/Lake/Content/Decomposition.htm)
Decomposers, which include bacteria, fungi, and other microorganisms for an intricate part of the food web. Despite the fact that decomposers break down organic matter, they function as sinks for plant and animal waste where they recycle nutrients for photosynthesis. For instance, phosphorus and nitrogen are regenerated to be bio available in the forms of $PO_{4}^{-3}$ and $NH_{4}^{+}$. The proposed model chosen does not account for the flux of $CO_{2}$ through the decomposer and therefore, as can be seen from the image below, no $CO_{2}$ or $CH_{4}$. However, there has been accounted for the $O_{2}$ consumed in the system as in certain zones it might be in sparce quantities.

![Decomposing_bacteria_explained](Decomposing_bacteria_explained.jpg)

## Flow models

In order to summarise the above information and the flow of elements a two way system split is proposed where the flow of carbon and nitrogen are seperated into two seperate flow models. In short $CO_{2}$ is converted by plants into carbohydrates, in turn the Nitrogen fixing bacteria consumes carbohydrates and reproduces $CO_{2}$ which is then consumed by the nitrifying, denitrifying and decomposing bacteria. THe carbon and nitrogen flow model is presented below:

![Carbon_Model](Carbon_Model.jpg)

## NItrogen flow model


![Nitrogen_Modell](Nitrogen_Modell.jpg)

# Next step
Once preliminary models of each type organism is produced and simulated in a two phase system. Our main goal will be to improve the accuracy of these models where the inclusion of a simplified energy balance of each organisms metabolism will act as a general starting point for these improvements.

## Stoichiometry

## Bacterial Reactions

In [2]:
import numpy
from numpy import linalg
import numpy as np

This section is dedicated to show the method used to obtain and fix the stoichiometric ratios of each reaction as set out above. Each plant and bacterial reaction with correponding component balances will be set out below to illustrate how the stoichiometry of each component was obtained.
Firstly, the general overall reaction of each organims was determined based on the characteristics of each organism. Once, the general reaction is determined the coefficients for each component is solved by means of using linear algebra.

Note: **All reactant rate values were taken as a negative i.e ($-r_{Reactant A}$) whereas products were taken as $r_{Product B}$ this is based on a rate of consumption and rate of production. In the event that a negate rate is obtained. It signals that the reactant was merely a product in the reaction or vice versa.**

### Nitrogen Fixation

For information regarding the chemical kinetics of nitrogen fixing cyanobacteria that make use of autotrophic growth(Continuous Cultivation as a Method to Assess the Maximum Specific Growth Rate of Photosynthetic Organisms)[https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6811504/]

(further on stoichiometry for whole nitrogen cycle and plants)[https://sci-hub.tw/10.1016/j.aquaculture.2006.03.019]




Nitrogen fixation as an overall chemical reaction can be described by the reaction below:


$$ CH_2O +Y_{SN} N_2 + Y_{SO}O_2 \to Y_{SX} CH_{1.2}O_{0.4}N{0.22} + Y_{SC} CO_2 + Y_{SW} H_2 O $$

By spesifying the ratio betweem biomass produced and the amount of glucose fed, in addition to a basis of 1 cmol glucose, the matrix below can be solved. Take cognisance of the fact that for the subsequent sections either $1mol$ $CH_20$ / ${NO_{3}}^{-}$ or $NH_{4}^{+}$ was used as a basis. Furthermore, a ratio between biomass produced based on the amount of $CH_20$ / ${NO_{3}}^{-}$ or , $NH_{4}^{+}$ fed was used to specify the system.









In [3]:
Fixation = np.matrix([
     [-1,0,0,1,1,0],
     [-2,0,0,1.2,0,2],
     [-1,0,-2,0.4,2,1],
     [0,-2,0,0.22,0,0],
     [1,0,0,0,0,0],
     [1,0,0,-1.643,0,0] ]) ## ratio between biomass produced per cmol substrate fed found in literature as set out in rough notebook


Fixation_sol = np.matrix([0,0,0,0,1,0]).T
Fixation_stoichiometry = np.linalg.solve(Fixation,Fixation_sol)
r_glucose,r_N2,r_O2,r_X,r_CO2,r_H20 = Fixation_stoichiometry
print(Fixation_stoichiometry)

[[1.        ]
 [0.0669507 ]
 [0.330493  ]
 [0.60864273]
 [0.39135727]
 [0.63481436]]


Consequently, the stoichiometry describing **Nitrogen Fixation**- bacteria can be defined as:
 $$ CH_2O +0.067 N_2 +0.33 O_2 \to 0.608 CH_{1.2}O_{0.4}N{0.22} +0.39  CO_2 +0.63 H_2 O $$


### Nitrification
a General formula for nitrification can be set out as follow:
 
$$NH_{4}^{+} +Y_{SO} O_2 +Y_{SC} CO_2\to Y_{SX}  CH_{1.4}O_{0.4}N_{0.22} +Y_{SN} NO_{3}^{-} + Y_{SW} H_2 O $$

In [4]:
Nitrification = np.matrix([[0,0,-1,1,0,0],
                           [-4,0,-2,1.4,0,2],
                           [0,-2,-2,0.4,3,1],
                           [-1,0,0,0.2,1,0],
                           [1,0,0,0,0,0],
                           [-1,0,0,1.07,0,0]]) # In this case a ratio between the ammount of ammonium and biomass production was specified

Nitrification_sol = np.matrix([[0,0,0,0,1,0]]).T
Nitrification_stoichiometry = np.linalg.solve(Nitrification,Nitrification_sol)
print(Nitrification_stoichiometry)

[[1.        ]
 [1.61214953]
 [0.93457944]
 [0.93457944]
 [0.81308411]
 [2.28037383]]


Therefore, Nitrifying Bacteria and its coupled reactions can be represented as set out in the equation below: 
$$ NH_{4}^{+} +1.61 O_2 +0.93 CO_2\to0.93  CH_{1.4}O_{0.4}N_{0.22} +0.81 NO_{3}^{-} +2.28 H_2 O $$


###  Denitrification


As illucidated to above, denitrification is a complex process that involes the reduction of Nitrates in order to form either N_{2} or N_{2}O. Based on empirical data denitrification including the growth of the bacteria performing the function can be represented as follow(page 499)[https://sci-hub.tw/10.1002/biot.200600078]:
$$ 1.08 CH_{3}OH + NO_{3}^{-} + 0.24 H_{2}CO_3 \to 0.056 C_{5}H_{7}O_{2}N +0.47N_2 +1.68H_{2}O +HCO_{3}^{-} $$
However, our simplified model has assumed that for now this ecosystem will operate in the absence of carbonic acid as a proton donor and therefore no bicarbonate will be produced:

$$CH_2O +Y_{SN} NO_{3}^{-} \to Y_{SX}  CH_{1.2}O_{0.51}N_{0.2} +Y_{SC} CO_{2} + Y_{SN_{2}} N_2 +Y_{SH} H_{2}O $$



For an in depth discussion regarding the reaction kinetics for denitrifying bacteria (see)[https://sci-hub.tw/10.2175/193864708788735556]

Subsequently, the matrix of denitrification can be set out as follow:


In [5]:
Denitrification = np.matrix([[-1,0,-1,1,0,0],
                             [-2,0,0,1.2,0,2],
                             [-1,-3,-2,0.51,0,1],
                             [0,-1,0,0.2,2,0],
                             [1,0,0,0,0,0],
                             [1,0,0,-3.86,0,0]]) # 

#In this instance a literature value relating the ratio 
#between carbon substrate fed to the amount of biomass 
#produced was used as a specification found in literature


Denitrification_sol = np.matrix([0,0,0,0,1,0]).T
Denitrification_stoichiometry = np.linalg.solve(Denitrification,Denitrification_sol)
print(Denitrification_stoichiometry)

[[ 1.        ]
 [ 0.48618307]
 [-0.74093264]
 [ 0.25906736]
 [ 0.2171848 ]
 [ 0.84455959]]


Therefore the stoichiometry as calculated translates to the following overall equation:
$$CH_2O + 0.48 NO_{3}^{-} \to0.26  CH_{1.2}O_{0.51}N_{0.2} +0.74 CO_{2} +0.22 N_2 +0.84 H_{2}O $$

### Decomposition:



For a starting model an assumption will be made that all biomass will just be returned to the basic components fed. Accordingly, all plants and bacteria will decompose into $NH_{4}^{+}$ and $CO_2$ at a spesific rate. It follows that the stoiciometry will be as follow:
$$ Biomass + O_2 \to NH_{4}^{+} +CO_{2}$$

[https://academic.oup.com/femsec/article/11/1/1/666725]
The value for $k_{d}$, the decay cofficient is given between $0.0048-0.0089 h^{-1}$ for plant matter(decay coefficient given in this article)[https://sci-hub.tw/10.4025/actascibiolsci.v34i4.9396]

# Plant reactions

In [6]:
import numpy as np

The as the sugars produced by photosynthesis will not only serve to feed the growth of the plant it is proposed that there be a model for photosynthesis as well as a growth equation for the plant in the system:
<ol> <li> Photsynthesis: $CO_{2} + H_{2}O \to CH_{2}O + O_{2}$ <li> Growth: $CH_2O +Y_{SO} O_2 + Y_{SN} NO_{3}^{-1}  \to Y_{SX} CH_{1.44}O_{0.66}N{0.04} +Y_{SC}  CO_2 +Y_{SW} H_2 O$ </ol>

Using the same logic above: using 1 cmol glucose as a basis and a specification ratio for biomass produced per cmol glucose fed as 


In [7]:
plantgrowth = np.matrix([[-1,0,0,1,1,0],
                         [-2,0,0,1.44,0,2],
                         [-1,-2,-3,0.66,2,1],
                         [0,0,-1,0.04,0,0],
                         [1,0,0,0,0,0],
                         [1,0,0,-3,0,0]]) 
plantgrowth_sol = np.matrix([0,0,0,0,1,0]).T
plantgrowth_stoichiometry = np.linalg.solve(plantgrowth,plantgrowth_sol)
print(plantgrowth_stoichiometry)


[[1.        ]
 [0.63666667]
 [0.01333333]
 [0.33333333]
 [0.66666667]
 [0.76      ]]


Therfore Plant growth can be represented as:
$$CH_2O +0.64 O_2 +0.01 NO_{3}^{-1}  \to0.33 CH_{1.44}O_{0.66}N{0.04} +0.67  CO_2 +0.76 H_2 O$$

## Stoichiometry summary:

The stoichiometry can be set out as follow:
<ol> 
<li> Fixation: $ CH_2O +0.067 N_2 +0.33 O_2 \to 0.608 CH_{1.2}O_{0.4}N{0.22} +0.39  CO_2 +0.63 H_2 O $ <li> Nitrification: $ NH_{4}^{+} +1.61 O_2 +0.93 CO_2\to0.93  CH_{1.4}O_{0.4}N_{0.22} +0.81 NO_{3}^{-} +2.28 H_2 O $ <li> Denitrification: $CH_2O + 0.48 NO_{3}^{-} \to0.26  CH_{1.2}O_{0.51}N_{0.2} +0.74 CO_{2} +0.22 N_2 +0.84 H_{2}O $ <li> Decomposition: $ X_{i} + O_2 \to NH_{4}^{+} +CO_{2}$ <li> Photosynthesis:  $CO_{2} + H_{2}O \to CH_{2}O + O_{2}$ <li>Plant growth: $CH_2O +0.64 O_2 +0.01 NO_{3}^{-1}  \to0.33 CH_{1.44}O_{0.66}N{0.04} +0.67  CO_2 +0.76 H_2 O$ </ol>

## Concentration profile model

As the stoichiometry is now fixed a set of differential equations can be used to solve  the concentration profiles of each chemical and biological specicies respectively. For starters $\mu$ will be assumed to approximately similar to that of $\mu_{max}$. However, as more is obtained these simplifying assumptions will dissipate and substrate inhibition as exepmlified by the Monod function will be used to determine the kinetics of each reaction.
The respective $\mu_{X_{n}}$ value will serve as a reaction rate. Using the $\mu$ value in conjunction with the stoichiometry as solved above the rate of production of each chemical and biological species is defined and by implication the system is solveable by means of numerical integration.

## Reaction rate constants and specific growth rate values:



<ol>
<li> $ k_{X_{Fix}} =\mu_{Fixation} = 0.05 h^{-1}$ 
    <li> $ k_{X_{Nit}} =\mu_{Nitrification} =0.034 h^{-1}$ 
    <li> $ k_{X_{Den}} =\mu_{Denitrification} =0.022 h^{-1}$
    <li> $ k_{Decomposition} =0.034 h^{-1}$ 
    <li> $k_{X_{Plant}}=\mu_{Plant} =0.1 h^{-1}$
    <li> $k_{gluc} = r_{CO_{2}} = 0.73 mol CO_{2}.h^{-1}$ </ol>
Knowing the various rates of each reaction enables the calculation of the concentration profile for each species in the ecosystem as they are related by the stoichiometry of each individual reaction containing them. 


      

## Individual reaction rates:

<ol> 
<li>$r_{Fix} = k_{X_{Fix}}\times C_{Fix}$
<li>$r_{Nit} = k_{X_{Nit}}\times C_{Nit}$
<li>$r_{Den} = k_{X_{Den}}\times C_{Den}$
<li>$r_{Dec} = k_{D}\times C_{Dec}$
<li>$r_{photosynthesis} = -r_{CO_{2}}\times C_{CO_{2}}$
<li>$r_{Plant} = \mu_{Plant}\times C_{Plant}$
 </ol>
 Accordingly, with these specification of each reaction rate, the rate of formation and consumption of each component in the system can be calculated using the stoichiometry in each reach raction and the total of each individual components' rate will be the rate of change of concentration of that component.

## Differential equations


<ol> $$\frac{dC_{CO2}}{dt}= r_{CO_{2}} C_{CO_{2}}$$
<li>$$\frac{dC_{CH_{2}O}}{dt}= r_{CH_{2}O} C_{CH_{2}O}$$
<li>$$\frac{dC_{O_{2}}}{dt}= r_{O_{2}} C_{O_{2}}$$
<li>$$\frac{dC_{H_{2}O}}{dt}= r_{H_{2}O} C_{H_{2}O}$$
<li>$$\frac{dC_{NH_{4}^{+}}}{dt}= r_{NH_{4}^{+}} C_{NH_{4}^{+}}$$
<li>$$\frac{dC_{NO_{3}^{-}}}{dt}= r_{NO_{3}^{-}} C_{NO_{3}^{-}}$$
     
<li>$$\frac{dC_{Fixation}}{dt}= k_{X_{Fix}} C_{X_{Fix}}$$
<li>$$\frac{dC_{Nitrification}}{dt}= k_{X_{Nit}} C_{X_{Nit}}$$
<li>$$\frac{dC_{Denitrification}}{dt}= k_{X_{Den}} C_{X_{Den}}$$


<li>$$\frac{dC_{Plant}}{dt}= k_{X_{Plant}} C_{X_{Plant}}$$ </ol>

As the decomposer is currently not modelled by means of a decomposing fungi/bacteria it will be assumed that there is no biomass and that each bacteria and respective bacteria in the ecosystem will disintegrate and form into the basic components of $NH_{4}^{+}$ and $CO_{2}$. Accordingly there is spesific differential equation for the decomposer and the decay constant will feature in both the former components' balance.